aws_signing_helper/tpm_signer.go

```diff
@@ -0,0 +1,345 @@
+	var paths []string
+	tpmdev := os.Getenv("TPM_DEVICE")
+	if tpmdev != "" {
+		paths = append(paths, tpmdev)
+	}
+	rw, err := tpm2.OpenTPM(paths...)
+	if err != nil {
+		return nil, err
+	}
+	defer rw.Close()
+
+	parentHandle := tpmutil.Handle(tpmv2Signer.tpmData.Parent)
+	if !handleIsPersistent(tpmv2Signer.tpmData.Parent) {
+		parentHandle, _, err = tpm2.CreatePrimary(rw, tpmutil.Handle(tpmv2Signer.tpmData.Parent), tpm2.PCRSelection{}, "", "", primaryParams)
+		if err != nil {
+			return nil, err
+		}
+		defer tpm2.FlushContext(rw, parentHandle)
+	}
+
+	keyHandle, _, err := tpm2.Load(rw, parentHandle, "", tpmv2Signer.tpmData.Pubkey[2:], tpmv2Signer.tpmData.Privkey[2:])
+	if err != nil {
+		return nil, err
+	}
+	defer tpm2.FlushContext(rw, keyHandle)
+
+	var algo tpm2.Algorithm
+	var shadigest []byte
+	var asn1Prefix []byte
+
+	switch opts.HashFunc() {
+	case crypto.SHA256:
+		sha256digest := sha256.Sum256(digest)
+		shadigest = sha256digest[:]
+		algo = tpm2.AlgSHA256
+		asn1Prefix = []byte{0x30, 0x31, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01, 0x05, 0x00, 0x04, 0x20}
+	case crypto.SHA384:
+		sha384digest := sha512.Sum384(digest)
+		shadigest = sha384digest[:]
+		algo = tpm2.AlgSHA384
+		asn1Prefix = []byte{0x30, 0x41, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x02, 0x05, 0x00, 0x04, 0x30}
+	case crypto.SHA512:
+		sha512digest := sha512.Sum512(digest)
+		shadigest = sha512digest[:]
+		algo = tpm2.AlgSHA512
+		asn1Prefix = []byte{0x30, 0x51, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x03, 0x05, 0x00, 0x04, 0x40}
+	}
+
+	if tpmv2Signer.public.Type == tpm2.AlgECC {
+
+		// For an EC key we lie to the TPM about what the hash is.
+		// It doesn't actually matter what the original digest was;
+		// the algo we feed to the TPM is *purely* based on the
+		// size of the curve itself. We truncate the actual digest,
+		// or pad with zeroes, to the byte size of the key.
+		pubKey, err := tpmv2Signer.public.Key()
+		if err != nil {
+			return nil, err
+		}
+		ecPubKey, ok := pubKey.(*ecdsa.PublicKey)
+		if !ok {
+			return nil, errors.New("Failed to obtain ecdsa.PublicKey")
+		}
+		bitSize := ecPubKey.Curve.Params().BitSize
+		byteSize := (bitSize + 7) / 8
+		if byteSize > sha512.Size {
+			byteSize = sha512.Size
+		}
+		switch byteSize {
+		case sha512.Size:
+			algo = tpm2.AlgSHA512
+		case sha512.Size384:
+			algo = tpm2.AlgSHA384
+		case sha512.Size256:
+			algo = tpm2.AlgSHA256
+		case sha1.Size:
+			algo = tpm2.AlgSHA1
+		default:
+			return nil, errors.New("Unsupported curve")
+		}
+
+		if len(shadigest) > byteSize {
+			shadigest = shadigest[:byteSize]
+		}
+
+		for len(shadigest) < byteSize {
+			shadigest = append([]byte{0}, shadigest...)
+		}
+
+		sig, err := tpm2.Sign(rw, keyHandle, "", shadigest, nil,
+			&tpm2.SigScheme{Alg: tpm2.AlgECDSA, Hash: algo})
+		if err != nil {
+			return nil, err
+		}
+		signature, err = asn1.Marshal(struct {
```

Comment:
Unclear whether it's worth doing, but we could leverage the `encodeEcdsaSigValue` function defined within `signer.go` here. But we would have to concatenate the two byte slices before feeding it into the function, making it potentially not worth it.

Reply:
Yeah, I think I tried that once and preferred it this way.